include/cppoptlib/solver/cmaesbsolver.h

```diff
@@ -0,0 +1,231 @@
+                      xk.coeffRef(d) = objFunc.lowerBound().coeffRef(d);
+                  } else if (xk.coeffRef(d) > objFunc.upperBound().coeffRef(d)) {
+                      dist = objFunc.upperBound().coeffRef(d) - xk.coeffRef(d);
+                      xk.coeffRef(d) = objFunc.upperBound().coeffRef(d);
+                  }
+                  penalty += (dist*dist) / eta;
+              }
+              costs[k] = objFunc(xk) + penalty/n;
+            }
+
+            if (Super::m_debug >= DebugLevel::High) {
+                std::cout << "arz" << std::endl << arz << std::endl;
+                std::cout << "arx" << std::endl << arx << std::endl;
+                std::cout << "costs " << costs.transpose() << std::endl;
+            }
+
+            std::vector<size_t> indices = index_partial_sort(costs, mu);
+            xmean = TVector::Zero(n);
+            zmean = TVector::Zero(n);
+            for (int k = 0; k < mu; k++) {
+              zmean += w[k]*arz.col(indices[k]);
+              xmean += w[k]*arx.col(indices[k]);
+            }
+            // Update constraint weights
+            for (int d = 0; d < n; d++) {
+                if (xmean.coeffRef(d) < objFunc.lowerBound().coeffRef(d) || xmean.coeffRef(d) > objFunc.upperBound().coeffRef(d)) {
+                    gamma.coeffRef(d) *= pow(1.1, std::max<Scalar>(1, mu_eff / (10*n)));
+                }
+            }
+            // Update evolution paths
+            ps = (1. - cs)*ps + sqrt(cs*(2. - cs)*mu_eff) * B*zmean;
+            Scalar hsig = (ps.norm()/sqrt(pow(1 - (1. - cs), (2 * (this->m_current.iterations + 1)))) < hsig_thr) ? 1.0 : 0.0;
+            pc = (1. - cc)*pc + hsig*sqrt(cc*(2. - cc)*mu_eff)*(B*D*zmean);
+            // Adapt covariance matrix
+            C = (1 - c1 - cmu)*C
+                + c1*(pc*pc.transpose() + (1 - hsig)*cc*(2 - cc)*C);
+            for (int k = 0; k < mu; ++k) {
+                TVector temp = B*D*arz.col(k);
+                C += cmu*w(k)*temp*temp.transpose();
+            }
+            sigma = sigma * exp((cs/ds) * ((ps).norm()/chi - 1.));
+
+            ++Super::m_current.iterations;
+            if ((Super::m_current.iterations - eigen_last_eval) == eigen_next_eval) {
+                // Update B and D
+                eigen_last_eval = Super::m_current.iterations;
+                Eigen::SelfAdjointEigenSolver<THessian> eigenSolver(C);
+                B = eigenSolver.eigenvectors();
+                D.diagonal() = eigenSolver.eigenvalues().array().sqrt();
+                if (Super::m_debug >= DebugLevel::High) {
+                    std::cout << "Updated hessian." << std::endl;
+                    std::cout << "C" << std::endl << C << std::endl;
+                    std::cout << "B" << std::endl << B << std::endl;
+                    std::cout << "D" << std::endl << D << std::endl;
+                }
+            }
+            Super::m_current.condition = D.diagonal().maxCoeff() / D.diagonal().minCoeff();
```

Comment:
It's defined using the C++11 `using` syntax on line 17. I think I was having trouble accessing members of the parent class because inheritance with templates doesn't work the same as for non-template classes. A similar pattern is used in ITK with `typedef`s